fxa-client/src/lib.rs

```diff
@@ -370,46 +371,39 @@ impl FirefoxAccount {
             Some(oauth_flow) => oauth_flow,
             None => return Err(ErrorKind::UnknownOAuthState.into()),
         };
-        self.handle_oauth_token_response(resp, oauth_flow.scoped_keys_flow)
-    }
-
-    fn handle_oauth_token_response(
-        &mut self,
-        resp: OAuthTokenResponse,
-        scoped_keys_flow: Option<ScopedKeysFlow>,
-    ) -> Result<OAuthInfo> {
-        let granted_scopes = resp.scope.split(" ").map(|s| s.to_string()).collect();
         // This assumes that if the server returns keys_jwe, the jwk argument is Some.
-        let keys = match resp.keys_jwe {
-            Some(jwe) => {
-                let scoped_keys_flow = scoped_keys_flow.expect(
+        match resp.keys_jwe {
+            Some(ref jwe) => {
+                let scoped_keys_flow = oauth_flow.scoped_keys_flow.expect(
                     "Insane state! If we are getting back a JWE this means we should have a JWK private key.",
                 );
-                Some(scoped_keys_flow.decrypt_keys_jwe(&jwe)?)
+                let decrypted_keys = scoped_keys_flow.decrypt_keys_jwe(jwe)?;
+                let scoped_keys: serde_json::Map<String, serde_json::Value> = serde_json::from_str(&decrypted_keys)?;
+                for (scope, key) in scoped_keys {
+                    self.state.scoped_keys.insert(scope.clone(), key.clone());
+                }
             }
             None => {
-                if scoped_keys_flow.is_some() {
+                if oauth_flow.scoped_keys_flow.is_some() {
                     error!("Expected to get keys back alongside the token but the server didn't send them.");
                     return Err(ErrorKind::TokenWithoutKeys.into());
-                } else {
-                    None
                 }
             }
         };
-        let since_epoch = SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .expect("Something is very wrong.");
-        let expires_at = since_epoch.as_secs() + resp.expires_in;
-        let oauth_info = OAuthInfo {
-            access_token: resp.access_token,
-            keys,
-            refresh_token: resp.refresh_token,
-            expires_at,
-            scopes: granted_scopes,
+        let refresh_token = match resp.refresh_token {
+            Some(ref refresh_token) => refresh_token.clone(),
+            None => return Err(ErrorKind::RefreshTokenNotPresent.into())
         };
-        self.oauth_cache_store(&oauth_info);
+        if let Some(ref old_refresh_token) = self.state.refresh_token {
+            let client = Client::new(&self.state.config);
+            client.destroy_oauth_token(&old_refresh_token.id).ok(); // ok() will ignore any error.
+        }
+        self.state.refresh_token = Some(RefreshToken {
+            id: refresh_token,
+            scopes: HashSet::from_iter(resp.scope.split(' ').map(|s| s.to_string())),
+        });
         self.maybe_call_persist_callback();
-        Ok(oauth_info)
+        Ok(())
```

Comment:
In this version, after the flow completion, reliers need to call get_oauth_token to get the keys? (does this make it a 3 step process instead of 2? begin, complete, get token)?

Reply:
Yup, I think that's an ok tradeoff as it only happens once.

Reply:
From discussion: let's destroy the access token we get.